File: LeetCode/add-binary.cpp
// https://leetcode.com/problems/add-binary/

class Solution
{
public:
	string addBinary(string a, string b)
	{
		string result = "";
		int aSize = a.size() - 1;
		int bSize = b.size() - 1;
		int binarySum = 0;
		int carry = 0;

		while (aSize >= 0 || bSize >= 0)
		{
			binarySum = carry;

			if (aSize >= 0)
			{
				binarySum += a[aSize] - '0';
			}
			if (bSize >= 0)
			{
				binarySum += b[bSize] - '0';
			}

			result += to_string(binarySum % 2);
			aSize--;
			bSize--;

			if (binarySum > 1)
			{
				carry = 1;
			}
			else
			{
				carry = 0;
			}
		}

		if (carry)
		{
			result += to_string(carry);
		}

		reverse(result.begin(), result.end());

		return result;
	}
};

File: LeetCode/add-binary.java

// @leet imports start
import java.util.*;
import java.math.*;
// @leet imports end

// @leet start
class Solution {
  public String addBinary(String a, String b) {
    int pointerA = a.length() - 1;
    int pointerB = b.length() - 1;
    int carry = 0;
    int result;

    StringBuilder stringBuilder = new StringBuilder();

    while (pointerA >= 0 || pointerB >= 0) {
      result = carry;

      if (pointerA >= 0) {
        result += a.charAt(pointerA) - '0';
      }

      if (pointerB >= 0) {
        result += (b.charAt(pointerB) - '0');
      }

      stringBuilder.append(result % 2);
      carry = result / 2;

      pointerA--;
      pointerB--;
    }

    if (carry != 0) {
      stringBuilder.append(carry);
    }

    return stringBuilder.reverse().toString();
  }
}
// @leet end
